libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/ReasoningBasedSimplifier.h>
+
+#include <libyul/optimiser/SSAValueTracker.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/AsmData.h>
+#include <libyul/Utilities.h>
+#include <libyul/Dialect.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libsmtutil/SMTPortfolio.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/CommonData.h>
+
+#include <utility>
+#include <memory>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::smtutil;
+
+void ReasoningBasedSimplifier::run(OptimiserStepContext& _context, Block& _ast)
+{
+	set<YulString> ssaVars = SSAValueTracker::ssaVariables(_ast);
+	ReasoningBasedSimplifier{_context.dialect, ssaVars}(_ast);
+}
+
+std::optional<string> ReasoningBasedSimplifier::invalidInCurrentEnvironment()
+{
+	// SMTLib2 interface is always available, but we would like to have synchronous answers.
+	if (smtutil::SMTPortfolio{}.solvers() <= 1)
+		return string{"No SMT solvers available."};
+	else
+		return nullopt;
+}
+
+void ReasoningBasedSimplifier::operator()(VariableDeclaration& _varDecl)
+{
+	if (_varDecl.variables.size() != 1 || !_varDecl.value)
+		return;
+	YulString varName = _varDecl.variables.front().name;
+	if (!m_ssaVariables.count(varName))
+		return;
+	m_variables.insert({varName, m_solver->newVariable("yul_" + varName.str(), defaultSort())});
```

Comment:
Why not just
```suggestion
	m_variables[varName] = m_solver->newVariable("yul_" + varName.str(), defaultSort());
```
? (which IMO is much easier on the eye)
Or do we want to assert that it wasn't there already?
```suggestion
	yulAssert(m_variables.insert({varName, m_solver->newVariable("yul_" + varName.str(), defaultSort())}).first);
```

Reply:
insert is needed because `Expression` is not default-constructible.